### ISSUE TITLE
Add _PYPY to twisted.python.compat

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#8503]
|[<img alt="markrwilliams's avatar" src="https://avatars.githubusercontent.com/u/704247?s=50" width="50" height="50">](https://github.com/markrwilliams)| @markrwilliams reported|
|-|-|
|Trac ID|trac#8503|
|Type|enhancement|
|Created|2016-06-26 05:35:23Z|
|Branch|https://github.com/twisted/twisted/tree/8503-markrwilliams-pypy-indicator|

`twisted.python.compat` defines `_PY3` to ease writing version-dependent code.  It should also have an analogous `_PYPY` variable so that there's a standard way to guard runtime-dependent code 

[#5257](https://github.com/twisted/twisted/issues/5257) would have resulted in a variable with the same effect if not the same name, but is no longer relevant given that it aims to support Python 2.4 and 2.5!  Please close it regardless of this ticket's outcome.

<details><summary>Searchable metadata</summary>

```
trac-id__8503 8503
type__enhancement enhancement
reporter__markrwilliams markrwilliams
priority__normal normal
milestone__PyPy_support PyPy-support
branch__8503_markrwilliams_pypy_indicator 8503-markrwilliams-pypy-indicator
branch_author__ 
status__closed closed
resolution__fixed fixed
component__core core
keywords__pypy_tests pypy tests
time__1466919323003177 1466919323003177
changetime__1468717617831814 1468717617831814
version__None None
owner__markrwilliams markrwilliams

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="markrwilliams's avatar" src="https://avatars.githubusercontent.com/u/704247?s=50" width="50" height="50">](https://github.com/markrwilliams)<a name="note_1"></a>|@markrwilliams commented|
|-|-|

See https://github.com/twisted/twisted/pull/221

**Comment by twisted-trac** [comment]
|[<img alt="rodrigc's avatar" src="https://avatars.githubusercontent.com/u/1895943?s=50" width="50" height="50">](https://github.com/rodrigc)<a name="note_3"></a>|@rodrigc commented|
|-|-|

Doing final testing under:
https://github.com/twisted/twisted/pull/223

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_4"></a>|@glyph commented|
|-|-|

How much PyPy runtime-dependent code do we have?  The only thing I can see is `twisted.persisted.styles`, which sort of looks like a mistake (why can't we get GC diagnostic information on serialization on pypy? it has the relevant API...).  Can we possibly avoid adding this?

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_5"></a>|@glyph commented|
|-|-|

To clarify a bit: `if _PY3:` is the worst possible way to guard version-dependent code.  What you really want for durability is `if <some runtime distinction relevant to the code below, like whether 'bytes' has a 'format' method>:`, but unfortunately this becomes so tedious in some cases that `if _PY3:` is a preferable shorthand.

But as far as I know we should not be adding anything that is PyPy specific _at all_, and in the few places where we might need to do that, it's just to check on a missing library or something, which would be better guarded with an `ImportError`.  So I am a bit concerned that we're jumping straight to the worst possible compatibility pattern before explaining why we need it.  Not to mention the fact that calling code could simply check `platform.python_implementation() == 'PyPy'` itself in the rare cases where it's necessary, which lets us skip a dozen or so extra lines of code being added to Twisted...

**Comment by twisted-trac** [comment]
|[<img alt="rodrigc's avatar" src="https://avatars.githubusercontent.com/u/1895943?s=50" width="50" height="50">](https://github.com/rodrigc)<a name="note_6"></a>|@rodrigc commented|
|-|-|

@glyph if you do not want to add twisted.python.compat._PYPY that is fine by me.

Mark wanted to use it in places like this:
https://github.com/twisted/twisted/pull/222/commits/faa58ad24e164acaf42778beeeb3f588001ed237

We still have that problem that for code that needs to be inside
if platform.python_implementation == "PyPy", this will run into problems with codecov,
since there is not working PyPy buildbot that uploads coverage reports to codecov.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_7"></a>|@glyph commented|
|-|-|

Replying to [rodrigc](#note_6):
> @glyph if you do not want to add twisted.python.compat._PYPY that is fine by me.

I would prefer to avoid it, but I think we need to discuss specific cases to see if we _can_ avoid it.

> Mark wanted to use it in places like this:
> https://github.com/twisted/twisted/pull/222/commits/faa58ad24e164acaf42778beeeb3f588001ed237

Great, brass tacks :).

This specific example is a good example of what I mean about not doing runtime detection specifically.  The fact that python 3 specifically unifies pickle and cPickle was an explicit API decision, but the fact that PyPy does is mostly a question of expediency.  It is a decision which might be reversed.  Plus, it's a decision which a new runtime we might want to support (IronPython, Jython, Pyston, etc) might decide to do either way.

So what this code should be doing is:

```
#!py
import pickle
import cPickle
if cPickle.PicklingError is pickle.PicklingError:
   "..."
```

This way, it works correctly on all available runtimes, anticipates any new ones, and doesn't require the use of hacks like `_PY3` or `_PYPY` which may, in the future, mis-identify this interpreter feature in any case.  Maybe when we finally get a CPython 2.8 - oops sorry I mean 2.7.20 - it'll make this change as well for some reason.

I am still definitely open to being convinced that it's necessary for other cases though, if there are other examples.

> We still have that problem that for code that needs to be inside
> if platform.python_implementation == "PyPy", this will run into problems with codecov,
> since there is not working PyPy buildbot that uploads coverage reports to codecov. 

This is a separate issue - has it been filed somewhere already?

**Comment by twisted-trac** [comment]
|[<img alt="rodrigc's avatar" src="https://avatars.githubusercontent.com/u/1895943?s=50" width="50" height="50">](https://github.com/rodrigc)<a name="note_8"></a>|@rodrigc commented|
|-|-|

There is this ticket for PyPy coverage [ticket:8505] but adiroiban got blocked,
so no further progress has been made.

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_9"></a>|@adiroiban commented|
|-|-|

I am also -1 for _PYPY and if possible I would like to get rid of _PY3 :)

I think that `platform.python_implementation() == 'PyPy'` should be fine... it should not be people to just use a bad pattern... and if there is really no other way, then the 'PyPy' check can be done via the platform.

For the pickle import and the universal error, I suggest to move it in compat and use ImportError

I would argue that compat should be the only place in which PY3 of PYPY should be used :)

------------

regarding the coverate, we have a dedicated ticket... but I have never run anything in PYPY so I was just trying to start the travis-ci part and leave someone else to handle the ugly PYPY coverage error :)

**Comment by twisted-trac** [comment]
|[<img alt="markrwilliams's avatar" src="https://avatars.githubusercontent.com/u/704247?s=50" width="50" height="50">](https://github.com/markrwilliams)<a name="note_10"></a>|@markrwilliams commented|
|-|-|

I should have provided more context in the original description.  Sorry!

Tests fail on PyPy for three basic reasons:

1. Simple resource leaks.  See [#8569](https://github.com/twisted/twisted/issues/8569) and [#8573](https://github.com/twisted/twisted/issues/8573).
2. Other invalid assumptions about when the interpreter collects garbage.  I haven't (yet) filed tickets for any particular instances of this.  Roughly, though, anything that expects `__del__` to run or weakrefs to expire by a particular point in time will eventually break on PyPy.
3. Apparently arbitrary implementation differences.  More on this shortly.

Clearly 1) doesn't require `_PYPY`.  2) probably doesn't; either tests that exercise this functionality are skipped in PyPy, in which case a single decorator containing the `platform.python_implementation` check suffices, or we figure out how to rewrite those tests to work on PyPy.  I'd like to do the first option, then the second.

3) is a different story.  My first comment on [#8504](https://github.com/twisted/twisted/issues/8504) describes the character of these issues. `cPickle` but not `pickle` lets you change the way functions are serialized, but no other builtin types, on CPython 2.7, while you can change how any are serialized PyPy.  Then there's [#8575](https://github.com/twisted/twisted/issues/8575): [twisted.python.rebuild](https://github.com/twisted/twisted/blob/trunk/twisted/python/rebuild.py#L105-L115) will fail with that `RebuildError` because [https://bitbucket.org/pypy/pypy/issues/2325/__slots__-and-__class__-assignment-is you can't reassign `__class__`] [https://bitbucket.org/pypy/pypy/src/36aba877dfa1fdfd695ea0fc44be494b7ef78d8e/pypy/objspace/std/test/test_typeobject.py?at=default&fileviewer=file-view-default#test_typeobject.py-876 when `__slots__` is involved.]   And there's [#8756](https://github.com/twisted/twisted/issues/8756): t.p.deprecate.deprecateModuleAttribute uses an unreliable heuristic to suppress warnings on import.  The list goes on; check the PyPy support milestone for more as they come up/trac lets me put them in.

Some of these problems are bugs in PyPy, some are bugs in Twisted, and still others are merely differences of opinion on the part of the implementers.  I don't doubt that over time these could be whittled down to a few issues that would readily succumb to an approach that's finer grained than `_PYPY`, but until then they'll keep the long tail of PyPy test failures alive.

I would prefer that we use a blunt instrument to cordone off problematic areas now, get the tests passing with informative skip messages, and then tackle that long tail over time.

Perhaps this approach is disingenuous.  It won't make PyPy a supported platform, but rather just a mostly supported platform.  But that's true of Python 3 right now, isn't it?  And we expect the builds not to fail for that.  For the same reasons `_PYPY` go in, and for the same reasons we should get to a place where we can toss it out!

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_11"></a>|@glyph commented|
|-|-|

OK, I think I'm sold; that makes sense.

Now we just have to get the buildbot working so we're not grinding our coverage numbers down while we fix this stuff :).

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_12"></a>|@adiroiban set owner to @markrwilliams|
|-|-|

OK :)

I can land it once test pass. @rodrigc when "working" on Twisted, is nice to be on #twisted-dev for quick communication with the other devs.

Beside adding _PYPY this branch also makes a change in styles without any comment about why we need that change.

Maybe it make perfect sense to make that change without any comment, but as a non PYPY user I have no idea why the PYPY implementation need this special handling.

I suggest that we should add a comment

**Comment by twisted-trac** [comment]
|<img alt="GitHub's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_13"></a>|GitHub set status to `closed`|
|-|-|

In changeset d9fc2ed9038a5f099ce8fe67283d08c3c364962a
```
#!CommitTicketReference repository="" revision="d9fc2ed9038a5f099ce8fe67283d08c3c364962a"
Merge markrwilliams/8503-markrwilliams-pypy-indicator: Add _PYPY helper.


Author: markrwilliams
Reviewer: adiroiban
Fixes: #8503
```
